### PR TITLE
add zero norm, inf norm support for p_norm op

### PR DESCRIPTION
**Pull request by ZHUI**
### PR types

Function optimization
### PR changes

OPs
### Describe

add zero norm, inf norm support for p_norm op

### CODE REVIEW

**Comment by ZHUI** [comment]
image

File: paddle/fluid/operators/p_norm_op.cc [hunk 1]
```diff
@@ -25,34 +25,49 @@ class PnormOpMaker : public framework::OpProtoAndCheckerMaker {
...
-        "(Tensor) Output tensor for the `(sum(x.pow(p)) + epsion).pow(1/p)`");
+    AddOutput("Out", "(Tensor) Output result tensor of p-norm");
     AddComment(R"DOC(
+Pnorm Operator.
+Given a tensor X, compute Lp-norm of X.
 
-Given a tensor, apply 2-normalization along the provided axis.
+When p = 0, defining $0^0 = 0$, the zero-norm of X is simply the number of non-zero elements of X.
+$$
+||X||_{0} = \lim_{p \rightarrow 0} \sum_i=1 |x_i|^p
```
Comment (ZHUI) [inline paddle/fluid/operators/p_norm_op.cc]: \sum_i=1  ->   \sum_i

File: paddle/fluid/operators/p_norm_op.cc [hunk 2]
```diff
@@ -63,24 +78,18 @@ class PnormOp : public framework::OperatorWithKernel {
...
     bool keepdim = ctx->Attrs().Get<bool>("keepdim");
+
     if (axis < 0) axis = xdim.size() + axis;
+    PADDLE_ENFORCE_GT(
+        xdim.size(), axis,
+        platform::errors::InvalidArgument(
+            "The input axis of p_norm is not support for "
+            "axis:%f >= xdim.size():%f, current input tensor X'shape is=[%s].",
+            axis, xdim.size(), xdim));
+
```
Comment (wawltor) [inline paddle/fluid/operators/p_norm_op.cc]: axis的区间 [-R , R)

File: paddle/fluid/operators/p_norm_op.cu [hunk 3]
```diff
@@ -49,20 +49,70 @@ __global__ void Pnorm(const T* x, const int pre,
...
+    int base = (i / post) * post * axis_n + (i % post);
+    T cur_min = inline_abs(x[base]);
+    for (int j = threadIdx.x; j < axis_n; j += blockDim.x) {
+      T x_ij_abs = inline_abs(x[base + j * post]);
+      if (cur_min > x_ij_abs) cur_min = x_ij_abs;
+    }
+    T reduce_result = BlockReduce(temp_storage).Reduce(cur_min, cub::Min());
+    if (threadIdx.x == 0) out_norm[i] = reduce_result;
   }
 }
```
Comment (wawltor) [inline paddle/fluid/operators/p_norm_op.cu:118]: 验证一下pytorch和paddle的inf，-inf，0的梯度是否一致

File: paddle/fluid/operators/p_norm_op.cc (hunk 1 above)
Comment (ZHUI) [reply paddle/fluid/operators/p_norm_op.cc]: done

File: paddle/fluid/operators/p_norm_op.cc (hunk 2 above)
Comment (ZHUI) [reply paddle/fluid/operators/p_norm_op.cc]: done

**Review by wawltor — APPROVED** [review]
LGTM

**Pull request merged**